NetSuite OpenAir (M4) connector guide (#428)
mulesoft/docs-connectors: modules/ROOT/pages/netsuite/netsuite-openair-connector.adoc

= NetSuite OpenAir Connector - Mule 4
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../../assets/images/

Support Category: https://www.mulesoft.com/legal/versioning-back-support-policy#anypoint-connectors[Select]

NetSuite OpenAir Connector v2.0

xref:release-notes::connector/netsuite-openair-connector-release-notes-mule-4.adoc[Release Notes]

> Anypoint Connector for NetSuite OpenAir (NetSuite OpenAir Connector) is a closed source enterprise connector that provides a connection between OpenAir and third-party software.

Reviewer: ```suggestion
Anypoint Connector for NetSuite OpenAir (NetSuite OpenAir Connector) is a closed-source enterprise connector that provides a connection between OpenAir and third-party software.
```

== Prerequisites

> This document assumes that you are familiar with NetSuite OpenAir, Mule, Anypoint Connectors, Anypoint Studio, Mule concepts, elements in a Mule flow, and Global Elements.

Reviewer: ```suggestion
To use this information, you must be familiar with NetSuite OpenAir, Mule runtime engine (Mule), Anypoint Connectors, Anypoint Studio, Mule concepts, elements in a Mule flow, and Global Elements.
```

To use this connector, you need:

* A NetSuite OpenAir instance

> * A set of valid credentials, including username, password, companyID, API namespace and key.

Reviewer: ```suggestion
* A set of valid credentials, including username, password, companyID, API namespace, and key.
```

* A MuleSoft Enterprise License

OpenAir complex types are limited to a set of OpenAir operations.

> See the OpenAir documentation for object types. For example, object `oaActualCost` works with `read`, `add`, `modify`, and `upsert` methods, while object `oaApproval` only works with the `submit` method.

Reviewer: ```suggestion
See the OpenAir documentation for information about its object types. For example, object `oaActualCost` works with `read`, `add`, `modify`, and `upsert` methods, while object `oaApproval` works only with the `submit` method.
```

[%header%autowidth.spread]
|===
|Application/Service|Version
|Mule|EE 4.1.0 and later
|Anypoint Studio|7.1 and later
|Java|JDK 8
|===

== Connect in Flow Designer

Flow Designer provides input source (triggers) for:

> 

Reviewer: Please reformat this list in the way that I have edited previous similar lists.
See the style reference.

You can use this connector in Anypoint Studio by adding it as a dependency in your Mule application.

== Add the Connector to Your Project

Anypoint Studio provides two ways to add the connector to your Studio project: from the 

> Exchange button in the Studio taskbar or from the Mule Palette.

Reviewer: Mule Palette view

Please make corrections from previous edits before submitting new text for editing.

=== Add the Connector Using Exchange

. In Studio, create a Mule project.
. Click the Exchange icon *(X)* in the upper-left of the Studio task bar.
. In Exchange, click *Login* and supply your Anypoint Platform username and password.
. In Exchange, search for "openair".
. Select the connector and click *Add to project*.
. Follow the prompts to install the connector.

=== Add the Connector in Studio

. In Studio, create a Mule project.
. In the Mule Palette view, click *(X) Search in Exchange*.
. In *Add Modules to Project*, type "openair" in the search field.
. Click this connector's name in *Available modules*.
. Click *Add*.
. Click *Finish*.

> == Apache Maven Pom File

Reviewer: POM file or .pom file

<dependency>
  <groupId>com.mulesoft.connectors</groupId>
  <artifactId>mule-netsuite-openair-connector</artifactId>
  <version>RELEASE</version>
  <classifier>mule-plugin</classifier>
</dependency>
----

Mule converts `RELEASE` to the latest release.

[[studioconfig]]
=== Configure in Studio

. Drag and drop the connector operation to the Studio Canvas (they are the same as are in Design Center).

> . Configure the global element for the connector:

Reviewer: global element is init. capped above, in prereqs: Global Element

+
image::netsuite/netsuite-openair-anypoint-config.png[anypoint config]
+
. If needed, configure a connection:
+
image::netsuite/netsuite-openair-anypoint-connect.png[anypoint connection]
+

> .. Enable the automatic reconnection feature:

Reviewer: Remove colon

.. Access the OpenAir Global Element configuration window from Studio.

> .. Click the *Advanced* (next to Connection) tab.

Reviewer: ```suggestion
.. Click the *Advanced* tab, next to Connection.
```

.. Select the *Standard* option in Reconnection Strategy.
.. Adjust the *Frequency (ms)* and *Reconnection Attempts* fields accordingly.

== Log HTTP Requests and Responses

To log the HTTP interactions of the connector with NetSuite OpenAir, configure the log4j2.xml file for the Mule app as follows, and use a Logger component to display the results:

[source,xml,linenums]
----
<AsyncLogger name="org.mule.service.http" level="DEBUG"/>
----

== Use Case: Add OpenAir Object

image::netsuite/netsuite-openair-flow-add.png[OpenAir use case flow]

. In Studio, drag an HTTP Listener operation onto the canvas and configure it to listen to host `0.0.0.0` at port `8081`.
. Drag the OpenAir Add operation connector into the flow and configure the connector as described in <<studioconfig>>.
. Drag a Transform Message component between the HTTP connector and the OpenAir connector.

> . In the Transform Message, insert the DataWeave script into the DataWeave editor to add oaCustomer objects to your NetSuite OpenAir instance:

Reviewer: ```suggestion
. In the Transform Message component, insert the DataWeave script into the DataWeave editor to add oaCustomer objects to your NetSuite OpenAir instance:
```

              company: "MuleSoft"
      },
      oaCustomer: {
        name: "Jane Doe",
        company: "Salesforce"
      }
    }
  }
}
----
+

> . Drag another Transform Message component after the OpenAir connector, and insert the DataWeave script into the Transform Message component's DataWeave editor. The script should return the IDs of the newly added OpenAir objects.

Reviewer: ```suggestion
. Drag another Transform Message component after the OpenAir connector, and insert the DataWeave script into the Transform Message component's DataWeave editor.
The script should return the IDs of the newly added OpenAir objects.
```
